File: designpattern/src/main/java/com/seniorlibs/designpattern/ch25v2/report/EmailReporter.java
package com.seniorlibs.designpattern.ch25v2.report;

import com.seniorlibs.designpattern.ch25v2.aggregate.Aggregator;
import com.seniorlibs.designpattern.ch25v2.model.RequestInfo;
import com.seniorlibs.designpattern.ch25v2.model.RequestStat;
import com.seniorlibs.designpattern.ch25v2.repository.MetricsStorage;
import com.seniorlibs.designpattern.ch25v2.viewer.StatViewer;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Timer;
import java.util.TimerTask;

/**
 * 组装类并定时触发执行统计显示
 */
public class EmailReporter {
    private static final Long DAY_HOURS_IN_SECONDS = 86400L;

    private MetricsStorage metricsStorage;
    private Aggregator aggregator;
    private StatViewer viewer;

    public EmailReporter(MetricsStorage metricsStorage, Aggregator aggregator, StatViewer viewer) {
        this.metricsStorage = metricsStorage;
        this.aggregator = aggregator;
        this.viewer = viewer;
    }

    public void startDailyReport() {
        Calendar calendar = Calendar.getInstance();
        calendar.add(Calendar.DATE, 1);
        calendar.set(Calendar.HOUR_OF_DAY, 0);
        calendar.set(Calendar.MINUTE, 0);
        calendar.set(Calendar.SECOND, 0);
        calendar.set(Calendar.MILLISECOND, 0);
        Date firstTime = calendar.getTime();
        Timer timer = new Timer();
        timer.schedule(new TimerTask() {
            @Override
            public void run() {
                long durationInMillis = DAY_HOURS_IN_SECONDS * 1000;
                long endTimeInMillis = System.currentTimeMillis();
                long startTimeInMillis = endTimeInMillis - durationInMillis;
                Map<String, List<RequestInfo>> requestInfos =
                        metricsStorage.getRequestInfos(startTimeInMillis, endTimeInMillis);
                Map<String, RequestStat> stats = aggregator.aggregate(requestInfos, durationInMillis);
                viewer.output(stats, startTimeInMillis, endTimeInMillis);
            }
        }, firstTime, DAY_HOURS_IN_SECONDS * 1000);
    }
}

File: designpattern/src/main/java/com/seniorlibs/designpattern/principle/frame/ch25v2/report/EmailReporter.kt
package com.seniorlibs.designpattern.principle.frame.ch25v2.report

import com.seniorlibs.designpattern.principle.frame.ch25v2.aggregate.Aggregator
import com.seniorlibs.designpattern.principle.frame.ch25v2.repository.MetricsStorage
import com.seniorlibs.designpattern.principle.frame.ch25v2.viewer.StatViewer
import java.util.*

/**
 * 组装类并定时触发执行统计显示
 */
class EmailReporter(
    private val metricsStorage: MetricsStorage,
    private val aggregator: Aggregator,
    private val viewer: StatViewer
) {
    fun startDailyReport() {
        val calendar = Calendar.getInstance()
        calendar.add(Calendar.DATE, 1)
        calendar[Calendar.HOUR_OF_DAY] = 0
        calendar[Calendar.MINUTE] = 0
        calendar[Calendar.SECOND] = 0
        calendar[Calendar.MILLISECOND] = 0
        val firstTime = calendar.time
        val timer = Timer()
        timer.schedule(object : TimerTask() {
            override fun run() {
                val durationInMillis = DAY_HOURS_IN_SECONDS * 1000
                val endTimeInMillis = System.currentTimeMillis()
                val startTimeInMillis = endTimeInMillis - durationInMillis
                val requestInfos =
                    metricsStorage.getRequestInfos(startTimeInMillis, endTimeInMillis)
                val stats = aggregator.aggregate(requestInfos, durationInMillis)
                viewer.output(stats, startTimeInMillis, endTimeInMillis)
            }
        }, firstTime, DAY_HOURS_IN_SECONDS * 1000)
    }

    companion object {
        private const val DAY_HOURS_IN_SECONDS = 86400L
    }
}